File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int currMin = 1, currMax = 1;
        int result = *max_element(nums.begin(), nums.end());
        
        for (int i=0; i<nums.size(); i++)
        {
            int temp = currMax * nums[i];
            currMax = max( currMax * nums[i], max( currMin * nums[i], nums[i]) );
            currMin = min( temp, min( currMin * nums[i], nums[i]) );
            result = max(result, currMax);
        }
        
        return result;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # Store the max Element
        res = max(nums)
        curMin, curMax = 1, 1
        
        for i in nums:
            temp = curMax * i
            curMax = max(temp, curMin * i, i)
            curMin = min(temp, curMin * i, i)
            res = max(res, curMax, curMin)
        
        return res